get session info and player profile

danthegoodman1 opened the pull request.

danthegoodman1 committed:
get session info

get player profile

danthegoodman1 commented:
Not sure if this is how we want to handle getting the info from the parent page

Hades32 commented on src/client.ts:
```diff
@@ -1,3 +1,81 @@
+    window.onmessage = (e) => {
+      // Only get this once
+      if (this.sessionToken === "") {
+        this.sessionToken = e.data
```
we should make the data a JSON object so we can extend it later

Hades32 commented on src/client.ts:
```diff
@@ -1,3 +1,81 @@
+    baseDomain?: string
+  }) {
+    // Register listener for iframe messaging
+    window.onmessage = (e) => {
```
The sender can't know if its message got received. We therefore should also add some kind of "ready" message back to the parent.

Hades32 commented on src/client.ts:
```diff
@@ -1,3 +1,81 @@
+      server_address: decoded.addr
+    }
+  }
+
```
missing a `gameFinished` method that notifies the parent window

danthegoodman1 committed:
convert data to JSON

add confirmation of getting session token

game over method with spin until confirmation

Hades32 commented on src/client.ts:
```diff
@@ -1,3 +1,106 @@
+   * Tells the Ultimate Arcade parent window that the game session has ended, and the game client can be removed from the page and the results shown.
+   */
+  async gameOver(): Promise<void> {
+    while (!this.gotEndResponse) {
+      window.top?.postMessage({msg: 'gameOver'}, '*')
+
+      // Sleep 50ms before checking again
+      await new Promise((r) => {
+        setTimeout(() => {
+          r(null)
+        }, 50)
+      })
```
why so complicated? The parent window is always there and can register its handlers before the iframe loads, so no reason to complicate things with retries here

danthegoodman1 replied:
Alright, but I generally go for retries everywhere... better to have and not need than the inverse

Hades32 replied:
I'd agree if this was a network call, but it isn't. This might lead to quite some confusion down the line: E.g. we could receive old messages from another iframe and not properly distinguish between them.
Actually thinking about this, it's a bit hard to figure out just by looking at the origin property to what iframe these messages relate to. We might need to introduce some random ID for the onMessage handler to properly differentiate them

danthegoodman1 replied:
Is there any case where we would have multiple origins?

Hades32 replied:
At the same time? There _shouldn't_, but you never know when the actual unload happens or what minor bug the frontend introduces which might only hide, but not unload an iframe. But in general, every game will have a slightly different origin. (This could be another way of handling this as ports get reused very rarely by agones)

danthegoodman1 replied:
we could pass the token back in every message since that is unique to a game session?

danthegoodman1 replied:
Or do you mean PER message

Hades32 replied:
yeah, token per session would

danthegoodman1 replied:
alright then we can just use the session token, will include in the messages

danthegoodman1 replied:
@Hades32 added

danthegoodman1 committed:
remove retry for gameOver

danthegoodman1 committed:
include token in response for context of game session
